Caching original dataone sources

yemoski opened the pull request:
I wrote the base code to get the original sources for dataone into a dictionary for 120 seconds.
I was hoping to get a review on the caching of the dictionary. Thanks!

nein09 commented on app/routes.py:
```diff
@@ -98,3 +104,31 @@ def handle_exception(e):
     return render_template("500_generic.html", e=e), 500
+
+#caching all the dataone original sources in memory (only update the datasources after 120 secs)
+@cache.cached(timeout=120, key_prefix='cache_original sources for dataone')
```
does this do it *every* 120 seconds, or 120 seconds after startup?

yemoski replied:
from the documentation it says it does it every 120 seconds, i can increase the time out interval if that would be okay

nein09 commented on app/routes.py:
```diff
@@ -13,6 +16,9 @@
 
 logger = logging.getLogger('app')
 
+cache = Cache()
```
This (and the method you write) makes sense, but I think it might make more sense to put it in https://github.com/nein09/polder-federated-search/blob/main/app/__init__.py - this file is more about handling requests that come in to the server and serving responses to them, and `__init__.py` does the setup stuff for the Flask app.

yemoski commented:
I would work on pulling the original sources from gleaner and write tests for both of them.

yemoski commented:
i added tests for the dataone original sources.
I also fixed the authors showing as an array to showing each author in the list. 
I tried adding it to the other pull request but it was closed before I could add it so I added it here.
I am not sure i know how to query the original source for the repos in gleaner

nein09 commented on app/__init__.py:
```diff
@@ -29,7 +40,57 @@
+
+    #node = mydoc.getElementsByTagName('node')
+    root = tree.getroot()
+    print(len(root))
```
looks like you left a debug statement in here - if it's handy, you could make it a debug log, but if you don't need it, please remove it

nein09 commented on app/search/search.py:
```diff
@@ -20,6 +20,7 @@ def __init__(self, **kwargs):
         self.keywords = kwargs.pop('keywords', [])
         self.origin = kwargs.pop('origin', [])
         self.id = kwargs.pop('id')
+        self.original_dict = kwargs.pop('original_dict', None)
```
Please name `original_dict` something like `datasource`, - we're never going to remember what `original_dict` refers to in 2 months.

yemoski replied:
True that, i doubt i would i remember the name.

nein09 commented on app/search/dataone.py:
```diff
@@ -103,6 +105,7 @@ def combined_search(self, **kwargs):
         return self.execute_query(query, page_number)
 
     def convert_result(self, result):
+        global original_dict 
```
I'm not sure this needs to be global - you could declare it here and set it to none instead, so that it's always defined, and then take out the `else` on line 157.

nein09 commented on app/search/dataone.py:
```diff
@@ -147,11 +150,21 @@ def convert_result(self, result):
                 )
+        # passing the dictionary with the original data sources
+        self.data_source_key =  result.pop('datasource', '').lstrip("urn:node:")
+        if self.data_source_key in app.datasources:
```
It occurred to me after our conversation about how to do this that because you made the method itself cached, you can call `app.get_original_dataone_sources()` here and do away with the `app.datasources` and `app.gleaner_datasources` variables entirely. If you do that, then the first time you make a query will be slow, but the rest of them will be fast, because the datasources will be in the cache.

yemoski replied:
i just changed it to app.get_original_dataone_sources() and i get this error that it does not exist. 
And i am pretty sure it cached
![image](https://user-images.githubusercontent.com/45835101/179301003-13b00daf-0396-4385-ae05-7bce84bbbc76.png)

nein09 replied:
oh right - that makes sense, because the methods aren't defined in the`app` object. Sorry, in order to do it that way, we'd need to put them in a different file and import them from there.

I wanted to try this out, but I can't run it right now because there are some things missing from `requirements.txt`. What did you need to `pip install` in order to get this to work on your machine? If you add those to `requirements.txt`,  I'll be able to try it - and deployed environments will need that in order to work.

nein09 commented on app/tests/search/test_dataone.py:
```diff
@@ -305,8 +305,39 @@ def test_Bounding_box_geometry(self):
+
+    def test_original_datasource(self):
+
+        self.search.max_score = 10
```
I think you don't need `max_score` anymore (did the test fail without it? if so, that's a bug)

nein09 commented:
The gleaner original sources would come from the `prov` stuff that I sent you (on slack I think) - but that's ok, those can happen as a follow- on.
